app/classifier/tasks/text/index.jsx

```diff
@@ -0,0 +1,184 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import _ from 'lodash';
+
+import GenericTask from '../generic';
+import TextTaskEditor from './editor';
+import TextTaskSummary from './summary';
+
+const LINEHEIGHT = 22.5;
+const NOOP = Function.prototype;
+
+export default class TextTask extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.textInput = React.createRef();
+
+    this.debouncedUpdateAnnotation = _.debounce(this.updateAnnotation, 500).bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.handleResize = this.handleResize.bind(this);
+    this.setTagSelection = this.setTagSelection.bind(this);
+    this.updateAnnotation = this.updateAnnotation.bind(this);
+  }
+
+  state = {
+    rows: 1,
+    value: this.props.annotation.value
+  }
+
+  componentDidMount() {
+    if (this.props.autoFocus) {
+      this.textInput.current.focus();
+    }
+    this.handleResize();
+  }
+
+  componentDidUpdate() {
+    this.handleResize();
+  }
+
+  componentWillUnmount() {
+    this.updateAnnotation();
+  }
+
+  setTagSelection(e) {
+    const textTag = e.target.value;
+    const startTag = `[${textTag}]`;
+    const endTag = `[/${textTag}]`;
+    const textArea = this.textInput.current;
+    const textAreaValue = textArea.value;
+    const selectionStart = textArea.selectionStart;
+    const selectionEnd = textArea.selectionEnd;
+    const textBefore = textAreaValue.substring(0, selectionStart);
+    let textAfter;
+    let value;
+
+    if (selectionStart === selectionEnd) {
+      textAfter = textAreaValue.substring(selectionStart, textAreaValue.length);
+      value = textBefore + startTag + endTag + textAfter;
+    } else {
+      const textInBetween = textAreaValue.substring(selectionStart, selectionEnd);
+      textAfter = textAreaValue.substring(selectionEnd, textAreaValue.length);
+      value = textBefore + startTag + textInBetween + endTag + textAfter;
+    }
+
+    this.setState({ value }, () => {
+      if (this.textInput.current.focus) {
+        this.textInput.current.setSelectionRange((value.length - textAfter.length), (value.length - textAfter.length));
+        this.textInput.current.focus();
+      }
+    });
+
+    this.updateAnnotation();
+  }
+
+  handleChange() {
+    const value = this.textInput.current.value;
+    if (value < this.state.value) {
+      this.setState({ rows: 1, value }, () => { this.handleResize(); });
+    } else {
+      this.setState({ value }, () => { this.handleResize(); });
+    }
+
+    this.debouncedUpdateAnnotation();
+  }
+
+  handleResize() {
+    const oldRows = this.textInput.current.rows;
+    const newRows = Math.floor(this.textInput.current.scrollHeight / LINEHEIGHT);
+
+    if (newRows && (newRows !== oldRows)) {
+      this.setState({ rows: newRows });
+    }
+  }
+
+  updateAnnotation() {
+    if (this.textInput.current) {
```

Comment:
Minor point but if `this.textInput.current` is undefined, then something has gone wrong here. Any pending calls to `updateAnnotation` should be canceled when the component unmounts, and this method should probably error if we're trying to sync annotation values after the text input has been removed from the DOM.